### ISSUE TITLE
[BUG] Newly added media to server B overrides watched on server A

### ISSUE THREAD

**Issue by pannal** [ArabCoders/watchstate#329]
Docker image: a1f10ddeb4edcd52cee2e875ba408ee7fa1806b7ef0453fec39c591430b8b214"

**Describe the bug**
I've watched 6/6 episodes of TV shows X on server A. Server B is missing the whole show.
Server B adds the show and its episodes - on the next sync, after server B has the show, watchstate marks the show on server A as unwatched.

**Expected behavior**
Watched is synced to server B once it adds the show and server A has it watched with the same user.

**Screenshots**
Server A, Plex, has a watched status on TV show X:
```
--------------------------
Task: import (Started: Fri, 23:00:36 CEST)
Command: /opt/app/bin/console state:import -v
Exit Code: 0 (Ended: Fri, 23:01:03 CEST)
--------------------------

[2023-08-25T23:00:36+02:00] NOTICE: SYSTEM: Preloading MemoryMapper data.
[2023-08-25T23:00:37+02:00] NOTICE: SYSTEM: Preloading MemoryMapper data is complete.
[2023-08-25T23:00:37+02:00] NOTICE: SYSTEM: Importing [home_plex] METADATA & PLAY STATE changes.
[2023-08-25T23:00:37+02:00] NOTICE: Ignoring [home_plex] [Videos] Unsupported agent type.
[2023-08-25T23:00:40+02:00] NOTICE: SYSTEM: Importing [ff_plex] METADATA & PLAY STATE changes.
[2023-08-25T23:00:41+02:00] NOTICE: SYSTEM: Waiting on [39] requests.
[2023-08-25T23:01:01+02:00] NOTICE: MAPPER: [home_plex] Updated and marked [Some awesome TV series (year) - 01x006] as [played].
[2023-08-25T23:01:03+02:00] NOTICE: SYSTEM: Finished waiting on [39] requests.
[2023-08-25T23:01:03+02:00] NOTICE: SYSTEM: Found [1] updated objects.
┌─────────┬───────┬─────────┬────────┐
│ Type    │ Added │ Updated │ Failed │
├─────────┼───────┼─────────┼────────┤
│ Movie   │ 0     │ 0       │ 0      │
├─────────┼───────┼─────────┼────────┤
│ Episode │ 0     │ 1       │ 0      │
└─────────┴───────┴─────────┴────────┘

--------------------------
Task: export (Started: Fri, 23:30:06 CEST)
Command: /opt/app/bin/console state:export -v
Exit Code: 0 (Ended: Fri, 23:30:07 CEST)
--------------------------
```

Server B, Plex, sees a new TV show after it's been added, same user:
```
--------------------------
Task: export (Started: Sat, 00:30:41 CEST)
Command: /opt/app/bin/console state:export -v
Exit Code: 0 (Ended: Sat, 00:30:42 CEST)
--------------------------

[2023-08-26T00:30:41+02:00] NOTICE: DATABASE: Loading changed items since [2023-08-25 22:30:32 CEST].
[2023-08-26T00:30:41+02:00] NOTICE: SYSTEM: Using push mode for [2] backends and export mode for [0] backends.
[2023-08-26T00:30:41+02:00] NOTICE: Push mode start.
[2023-08-26T00:30:42+02:00] NOTICE: Push mode ends.
[2023-08-26T00:30:42+02:00] NOTICE: SYSTEM: Sending [6] change play state requests.
[2023-08-26T00:30:42+02:00] NOTICE: Marked [home_plex] [Some awesome TV series (year) - 01x001] as [Unplayed].
[2023-08-26T00:30:42+02:00] NOTICE: Marked [home_plex] [Some awesome TV series (year) - 01x002] as [Unplayed].
[2023-08-26T00:30:42+02:00] NOTICE: Marked [home_plex] [Some awesome TV series (year) - 01x003] as [Unplayed].
[2023-08-26T00:30:42+02:00] NOTICE: Marked [home_plex] [Some awesome TV series (year) - 01x004] as [Unplayed].
[2023-08-26T00:30:42+02:00] NOTICE: Marked [home_plex] [Some awesome TV series (year) - 01x005] as [Unplayed].
[2023-08-26T00:30:42+02:00] NOTICE: Marked [home_plex] [Some awesome TV series (year) - 01x006] as [Unplayed].
[2023-08-26T00:30:42+02:00] NOTICE: SYSTEM: Sent [6] change play state requests.
[2023-08-26T00:30:42+02:00] NOTICE: Using WatchState Version - '4bbe4ee'.
``` 

```
[ Basic Report ]

WatchState Version: 4bbe4ee
PHP Version: 8.2.9
Timezone: Europe/Berlin
Running in Container? Yes
Data Path: /config
Temp Path: /config
Database Migrated?: Yes
Has .env file? No
Is Tasks Runner working? Yes
Report Generated At: 2023-08-26T03:13:20+00:00

[ Backends ]

[ Plex ==> home_plex ]

Is backend URL HTTPS? Yes
Has Unique Identifier? Yes
Has User? Yes
Export Enabled? Yes
Time since last export? 2023-08-25T22:30:42+00:00
Play state import enabled? Yes
Metadata only import enabled? No
Time since last import? 2023-08-26T03:00:09+00:00
Has webhook token? Yes
Is webhook match user id enabled? Yes
Is webhook match backend unique id enabled? Yes
Has custom options? No
{}
[ Plex ==> ff_plex ]

Is backend URL HTTPS? Yes
Has Unique Identifier? Yes
Has User? Yes
Export Enabled? Yes
Time since last export? 2023-08-25T22:30:42+00:00
Play state import enabled? Yes
Metadata only import enabled? No
Time since last import? 2023-08-26T03:00:09+00:00
Has webhook token? Yes
Is webhook match user id enabled? Yes
Is webhook match backend unique id enabled? Yes
Has custom options? No
{}
[ Tasks ]

[ Import ]

Is Task enabled? Yes
Which flags are used to run the task? -v
When the task scheduled to run at? 0 */1 * * *
When is the next scheduled run? 2023-08-26T04:00:00+00:00

[ Export ]

Is Task enabled? Yes
Which flags are used to run the task? -v
When the task scheduled to run at? 30 */1 * * *
When is the next scheduled run? 2023-08-26T03:30:00+00:00

[ Push ]

Is Task enabled? No

[ Backup ]

Is Task enabled? No

[ Prune ]

Is Task enabled? Yes
Which flags are used to run the task? -v
When the task scheduled to run at? 0 */12 * * *
When is the next scheduled run? 2023-08-26T10:00:00+00:00

[ Indexes ]

Is Task enabled? Yes
Which flags are used to run the task? -v
When the task scheduled to run at? 0 3 * * 3
When is the next scheduled run? 2023-08-30T01:00:00+00:00

[ Requests ]

Is Task enabled? Yes
Which flags are used to run the task? -v --no-stats
When the task scheduled to run at? */2 * * * *
When is the next scheduled run? 2023-08-26T03:14:00+00:00

[ Logs ]

[ /config/logs/app.20230826.log ]

app log file is empty or does not exists.

[ /config/logs/access.20230826.log ]

access log file is empty or does not exists.

[ /config/logs/task.20230826.log ]

[2023-08-26T05:00:31+02:00] NOTICE: SYSTEM: Finished waiting on [39] requests.
[2023-08-26T05:00:31+02:00] NOTICE: SYSTEM: Found [6] updated objects.
┌─────────┬───────┬─────────┬────────┐
│ Type    │ Added │ Updated │ Failed │
├─────────┼───────┼─────────┼────────┤
│ Movie   │ 0     │ 0       │ 0      │
├─────────┼───────┼─────────┼────────┤
│ Episode │ 0     │ 6       │ 0      │
└─────────┴───────┴─────────┴────────┘

[ /config/logs/app.20230825.log ]

app log file is empty or does not exists.

[ /config/logs/access.20230825.log ]

access log file is empty or does not exists.

[ /config/logs/task.20230825.log ]

--------------------------
Task: export (Started: Fri, 23:30:06 CEST)
Command: /opt/app/bin/console state:export -v
Exit Code: 0 (Ended: Fri, 23:30:07 CEST)
--------------------------
[2023-08-25T23:30:07+02:00] NOTICE: DATABASE: Loading changed items since [2023-08-25 22:30:32 CEST].
[2023-08-25T23:30:07+02:00] NOTICE: DATABASE: No play state change detected since [2023-08-25 22:30:32 CEST].

```

**Comment by pannal** [comment]
docker compose:
```
    environment:
      - WS_TZ=Europe/Berlin # Set timezone.
      - WS_CRON_IMPORT=1
      - WS_CRON_EXPORT=1
```

**Comment by ArabCoders** [comment]
Thanks for your report, If i remember correctly, This is the intended behavior as it's the only we we can support mark as unwatched. 

However, It's not totally lost please do re-read the instructions in the `config:add` if you add new server that is not in sync with your current play state you are supposed to only enable the metadata import and force export your current watch state to the new server. Once they are in sync you can go ahead and enable import from the new server. 

Pleas try it and report back. If this doesn't work i can take closer look at the codebase again. 

In the meantime i will try to look for some safeguards to prevent this specific condition from occurring if possible.

**Comment by pannal** [comment]
> However, It's not totally lost please do re-read the instructions in the config:add if you add new server that is not in sync with your current play state you are supposed to only enable the metadata import and force export your current watch state to the new server.

Well the server was already added and synced. The second one just didn't have the TV show added, and it was added after I watched it on the other server. This use case has to be pretty common for people with two servers.

> In the meantime i will try to look for some safeguards to prevent this specific condition from occurring if possible.

Don't you get the info from the Plex Media Server about when the item in question was added? If an item was added later than its already known watched state, it shouldn't override it to unwatched, but instead set it to watched on the other server as well.

**Comment by ArabCoders** [comment]
Honestly, rechecking the code as it's has been long time seeing any problems related to this it seems i do have safeguards in place to prevent such case you can clearly see from this code

https://github.com/ArabCoders/watchstate/blob/4bbe4eed498508b8e3fafd712f62889ac39af2e6/src/Libs/Entity/StateEntity.php#L312

i only mark the item as unwatched if there are some conditions met, which involve checking the already has the item reported before and was marked as watched previously from said backend. By chance did you re-use the same backend name thus reaching some kind of edge case?

**Comment by pannal** [comment]
> By chance did you re-use the same backend name thus reaching some kind of edge case?

How do you mean? I've posted my config above, they have different names, or am I missing something?

**Comment by ArabCoders** [comment]
No worries, could you please re-run the commands again using the flags i gave to have test case so i can proceed, i tested right now with two Plex backends, and the expected behavior of not marking already marked as played items working when given backend didn't report video as played previously. However the new backend wont be marked as played as the date on video is newer than the marked as played item in database. to get it in sync i had to force push with ignore date flag set. That did the trick and set the new server status to be in sync with the database.

**Comment by pannal** [comment]
Umm, how exactly do I do that? Do you want me to delete the media on the second server again, re-add it, while watchstate runs with the parameters you gave?

**Comment by ArabCoders** [comment]
> > could you please re-run the commands again using the flags i gave
> 
> Umm, how exactly do I do that? Do you want me to delete the media on the second server again, re-add it, while watchstate runs with the parameters you gave?

Yes please, try to recreate the same situation with the flags turned on so we can trace the problem and where the problem lies, It will be great if you start with new database config as well just to make sure i can reproduce the edge case.

**Comment by pannal** [comment]
Pinged you on Discord

**Comment by ArabCoders** [comment]
As we discussed in discord, It seems the code is working correctly, The problem lies in the Memory Mapper handling of objects which has future date compared to db, which is what usually expected. I have an idea for fix in mind i'll update this issue once that fix is deployed in `:dev` branch

**Comment by ArabCoders** [comment]
I have a potential fix for the issue related to the mapper in the `dev` branch please switch to it and report back. Thank you

**Comment by ArabCoders** [comment]
This is fixed in #330 closing. Free feel to re-open if the problem still exists for you.
